src/view.rs

```diff
@@ -0,0 +1,598 @@
+    #[must_use]
+    fn port_index(&self, node: NodeIndex, offset: PortOffset) -> Option<PortIndex>;
+
+    /// Iterates over all the ports of the `node` in the given `direction`.
+    #[must_use]
+    fn ports(&self, node: NodeIndex, direction: Direction) -> Self::NodePorts<'_>;
+
+    /// Iterates over the input and output ports of the `node` in sequence.
+    #[must_use]
+    fn all_ports(&self, node: NodeIndex) -> Self::NodePorts<'_>;
+
+    /// Iterates over all the input ports of the `node`.
+    ///
+    /// Shorthand for [`PortView::ports`].
+    #[must_use]
+    #[inline]
+    fn inputs(&self, node: NodeIndex) -> Self::NodePorts<'_> {
+        self.ports(node, Direction::Incoming)
+    }
+
+    /// Iterates over all the output ports of the `node`.
+    ///
+    /// Shorthand for [`PortView::ports`].
+    #[must_use]
+    #[inline]
+    fn outputs(&self, node: NodeIndex) -> Self::NodePorts<'_> {
+        self.ports(node, Direction::Outgoing)
+    }
+
+    /// Returns the input port at the given offset in the `node`.
+    ///
+    /// Shorthand for [`PortView::port_index`].
+    #[must_use]
+    fn input(&self, node: NodeIndex, offset: usize) -> Option<PortIndex>;
+
+    /// Returns the output port at the given offset in the `node`.
+    ///
+    /// Shorthand for [`PortView::port_index`].
+    #[must_use]
+    fn output(&self, node: NodeIndex, offset: usize) -> Option<PortIndex>;
+
+    /// Returns the number of input ports of the `node`.
+    ///
+    /// Shorthand for [`PortView::num_ports`].
+    #[must_use]
+    #[inline]
+    fn num_inputs(&self, node: NodeIndex) -> usize {
+        self.num_ports(node, Direction::Incoming)
+    }
+
+    /// Returns the number of output ports of the `node`.
+    ///
+    /// Shorthand for [`PortView::num_ports`].
+    #[must_use]
+    #[inline]
+    fn num_outputs(&self, node: NodeIndex) -> usize {
+        self.num_ports(node, Direction::Outgoing)
+    }
+
+    /// Returns the number of ports of the `node` in the given `direction`.
+    #[must_use]
+    fn num_ports(&self, node: NodeIndex, direction: Direction) -> usize;
+
+    /// Iterates over all the port offsets of the `node` in the given `direction`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use portgraph::PortGraph;
+    /// # use portgraph::{Direction, PortOffset, PortView, LinkView};
+    /// let mut graph = PortGraph::new();
+    /// let node = graph.add_node(0, 2);
+    ///
+    /// assert!(graph.links(node, Direction::Incoming).eq([]));
+    /// assert!(graph.port_offsets(node, Direction::Outgoing).eq([PortOffset::new_outgoing(0), PortOffset::new_outgoing(1)]));
+    /// ```
+    #[must_use]
+    fn port_offsets(&self, node: NodeIndex, direction: Direction) -> Self::NodePortOffsets<'_>;
+
+    /// Iterates over the input and output port offsets of the `node` in sequence.
+    #[must_use]
+    fn all_port_offsets(&self, node: NodeIndex) -> Self::NodePortOffsets<'_>;
```

Comment:
can this not be a blanket impl? Just by chaining input_offsets and output_offsets

Reply:
That would be another iterator type, `Chain<PortOffsets, PortOffsets>`. In most cases we can do things more efficiently since internally the ports are contiguous.